pxd_fastpath.c

```diff
@@ -419,33 +421,43 @@ static void pxd_io_failover(struct work_struct *ws)
 	struct pxd_io_tracker *head = container_of(ws, struct pxd_io_tracker, wi);
 	struct pxd_device *pxd_dev = head->pxd_dev;
 	bool cleanup = false;
+	bool reroute = false;
+	int rc;
 
 	BUG_ON(head->magic != PXD_IOT_MAGIC);
 	BUG_ON(pxd_dev->magic != PXD_DEV_MAGIC);
 
 	spin_lock(&pxd_dev->fp.fail_lock);
-	if (pxd_dev->fp.fastpath &&
-		pxd_dev->fp.active_failover == PXD_FP_FAILOVER_NONE) {
-		pxd_dev->fp.active_failover = PXD_FP_FAILOVER_ACTIVE;
-		cleanup = true;
+	if (pxd_dev->fp.fastpath) {
+		if (pxd_dev->fp.active_failover == PXD_FP_FAILOVER_NONE) {
+			pxd_dev->fp.active_failover = PXD_FP_FAILOVER_ACTIVE;
+			cleanup = true;
+		}
+		__pxd_add2failQ(pxd_dev, head);
+	} else if (pxd_dev->fp.active_failover == PXD_FP_FAILOVER_NONE) {
+		reroute = true;
 	}
+
 	spin_unlock(&pxd_dev->fp.fail_lock);
 
 	if (cleanup) {
 		disableFastPath(pxd_dev, true);
-	}
-
-	if (!pxd_dev->connected || pxd_dev->removing) {
-		printk_ratelimited(KERN_ERR"%s: pxd%llu: px is disconnected, failing IO.\n", __func__, pxd_dev->dev_id);
-		BIO_ENDIO(head->orig, -ENXIO);
-	} else {
-		// switch to native path
+		rc = pxd_initiate_failover(pxd_dev);
+		// If userspace cannot be informed of a failover event, force abort all IO.
+		if (rc) {
```

Comment:
This would abort IOs if rc==EBUSY i.e we have a failover op in progress. I don't think that's correct ?

Reply:
failover gets initiated only once by the first failing IO. Others just get enqueued to the failQ.